File: DSA/Yashvardhan Guliyani/Day6_q2.java
class ListNode {
    int val;
    ListNode next;
    
    ListNode(int x) {
        val = x;
        next = null;
    }
}

public class LinkedListMergePoint {
    public static int findMergePoint(ListNode head1, ListNode head2) {
        if (head1 == null || head2 == null) {
            return -1; 
        }
        
        ListNode pointer1 = head1;
        ListNode pointer2 = head2;
        
        while (pointer1 != pointer2) {
            pointer1 = pointer1.next;
            pointer2 = pointer2.next;
            if (pointer1 == null) {
                pointer1 = head2;
            }
            if (pointer2 == null) {
                pointer2 = head1;
            }
        }
        return (pointer1 != null) ? pointer1.val : -1;
    }
    
    public static void main(String[] args) {
        ListNode head1 = new ListNode(1);
        head1.next = new ListNode(2);
        head1.next.next = new ListNode(3);
        ListNode head2 = new ListNode(4);
        head2.next = new ListNode(5);
        ListNode mergeNode = new ListNode(3);
        head1.next.next = mergeNode;
        head2.next.next = mergeNode;
        int mergeValue = findMergePoint(head1, head2);
        System.out.println("Merge point value: " + mergeValue);
    }
}

File: DSA/Hima-Swarupa-Padala/day6q2.py
class ListNode:
    def __init__(self, val):
        self.val = val
        self.next = None

def find_merge_point(head1, head2):
    node_dict = {}
    temp = head1
    while temp:
        node_dict[temp] = True
        temp = temp.next

    temp = head2
    while temp:
        if temp in node_dict:
            return temp.val
        temp = temp.next

    return None

a1 = ListNode(1)
a1.next = ListNode(2)
a1.next.next = ListNode(3)

a2 = ListNode(4)
a2.next = ListNode(5)
a2.next.next = a1.next.next  

merge_point = find_merge_point(a1, a2)
print(merge_point)  